ercc/registry/registry.go

```diff
@@ -0,0 +1,255 @@
+	if err != nil {
+		return nil, err
+	}
+
+	credentials, err := ctx.GetStub().GetState(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return credentials, nil
+}
+
+// Optional Post-MVP;
+// returns a list of all provisioned enclaves for a given chaincode id. A provisioned enclave is a registered enclave
+// that has also the chaincode decryption key.
+func (rs *Contract) QueryListProvisionedEnclaves(ctx contractapi.TransactionContextInterface, chaincodeId string) ([]string, error) {
+
+	iter, err := ctx.GetStub().GetStateByPartialCompositeKey("namespaces/credentials", []string{chaincodeId})
+	defer iter.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	var enclaveIds []string
+
+	for iter.HasNext() {
+		q, err := iter.Next()
+		if err != nil {
+			return nil, err
+		}
+
+		_, res, err := ctx.GetStub().SplitCompositeKey(q.Key)
+		if err != nil {
+			return nil, err
+		}
+
+		enclaveId := res[1]
+
+		// next check that for each enclaveID there also exists a CCKeyRegistrationMessage
+		k, err := ctx.GetStub().CreateCompositeKey("namespaces/provisioned", []string{chaincodeId, enclaveId})
+		if err != nil {
+			return nil, err
+		}
+
+		p, err := ctx.GetStub().GetState(k)
+		if err != nil {
+			return nil, err
+		}
+
+		if p != nil {
+			enclaveIds = append(enclaveIds, enclaveId)
+		}
+	}
+
+	return enclaveIds, nil
+}
+
+// returns the chaincode encryption key for a given chaincode id
+func (rs *Contract) QueryChaincodeEncryptionKey(ctx contractapi.TransactionContextInterface, chaincodeId string) ([]byte, error) {
+	//input chaincodeId string
+
+	//return chaincode_ek []byte
+	return nil, nil
+}
+
+// register a new FPC chaincode enclave instance
+func (rs *Contract) RegisterEnclave(ctx contractapi.TransactionContextInterface, credentialsBase64 string) error {
+	log.Printf("RegisterEnclave")
+
+	credentialBytes, _ := base64.StdEncoding.DecodeString(credentialsBase64)
+
+	var credentials protos.Credentials
+
+	if len(credentialBytes) == 0 {
+		return errors.New("credential message is empty")
+	}
+
+	if err := proto.Unmarshal(credentialBytes, &credentials); err != nil {
+		return errors.Wrap(err, "invalid credential bytes")
+	}
+
+	if credentials.SerializedAttestedData == nil {
+		return errors.New("attested data is empty")
+	}
+
+	if len(credentials.Evidence) == 0 {
+		return errors.New("evidence is empty")
+	}
+
+	// get attested data from credentials
+	var attestedData protos.Attested_Data
+	if err := ptypes.UnmarshalAny(credentials.SerializedAttestedData, &attestedData); err != nil {
+		return errors.Wrap(err, "invalid attested data message")
+	}
+
+	if err := checkAttestedData(ctx, rs.Verifier, rs.PEvaluator, &attestedData, &credentials); err != nil {
+		return err
+	}
+
+	chaincodeId := attestedData.CcParams.ChaincodeId
+	enclaveId := utils.GetEnclaveId(&attestedData)
+
+	key, err := ctx.GetStub().CreateCompositeKey("namespaces/credentials", []string{chaincodeId, enclaveId})
```

Comment:
Wouldn't you have to do here also some check whether there are already registered enclaves for that chaincode and, if so, verify that they are consistent?

Reply:
added todos